### PR TITLE
[AVRO] Prioritise Avro providers from "extensions/core"

### LINKED ISSUE #25601: [Failing Test]: Python PostCommit failing due to duplicate AvroSchemaIO autoservice
### What happened?

Python PostCommit jdbcio_xlang_it_test failing. Jenkins log does not show much message, but running locally the actual error shows:
```
ERROR:apache_beam.utils.subprocess_server:Starting job service with ['java', '-jar', '/Users/yathu/.apache_beam/cache/jars/composite-jars/9b6e2cb01bd723cbd87a5e71462f25664f165993697afc98e1c64c34bf814f98.jar', '50709', '--filesToStage=/Users/yathu/dev/virtualenv/py38beam/lib/sdks/java/extensions/schemaio-expansion-service/build/libs/beam-sdks-java-extensions-schemaio-expansion-service-2.47.0-SNAPSHOT.jar,/Users/yathu/.apache_beam/cache/jars/postgresql-42.2.16.jar']
ERROR:apache_beam.utils.subprocess_server:Error bringing up service
Traceback (most recent call last):
  File "/Users/yathu/dev/virtualenv/py38beam/lib/python3.8/site-packages/apache_beam/utils/subprocess_server.py", line 88, in start
    raise RuntimeError(
RuntimeError: Service failed to start up with error 1
Traceback (most recent call last):
```
This is because there are two `@AutoService AvroSchemaIOProvider` classes so the schemaio expansion service fails to start.

The fix could either be similar to the workaround here: https://github.com/apache/beam/blob/011296c14659f80c8ecbeefda79ecc3f1113bd95/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java#L45

Or simply remove AvroSchemaIOProvider in core. 

### Issue Failure

Failure: Test is continually failing

### Issue Priority

Priority: 1 (unhealthy code / failing or flaky postcommit so we cannot be sure the product is healthy)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by aromanenko-dev**
Fixes #25601

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
Run Python3.10 PostCommit

**Comment by aromanenko-dev** [comment]
CC: @mosche @Abacn @sclukas77

**Review by mosche — CHANGES_REQUESTED** [review]
@aromanenko-dev pls have a look, this changed the general behavior in unintended ways.

**Comment by aromanenko-dev** [comment]
Run Python 3.10 PostCommit

File: sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java [hunk 1]
```diff
@@ -50,22 +51,50 @@ public class ExternalSchemaIOTransformRegistrar implements ExternalTransformRegi
...
+      for (SchemaIOProvider provider : ServiceLoader.load(SchemaIOProvider.class)) {
+        // Avro provider is treated as a special case since two Avro providers may want to be loaded
+        // from "core" (deprecated) and from "extensions/avro" (actual) - but only one must succeed.
+        // TODO: we won't need this check once all Avro providers from "core" will be
+        // removed
+        if (provider.identifier().equals("avro")) {
+          // Avro provider from "extensions/avro" must have a priority.
+          if (provider.getClass().getName().startsWith("org.apache.beam.sdk.extensions.avro")) {
+            // Load Avro provider from "extensions/avro" by any case.
+            loadProvider(providers, provider);
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java]: Changed

File: sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java [hunk 2]
```diff
@@ -50,22 +53,55 @@ public class ExternalSchemaIOTransformRegistrar implements ExternalTransformRegi
...
+          if (provider.getClass().getName().startsWith("org.apache.beam.sdk.extensions.avro")) {
+            // Load Avro provider from "extensions/avro" by any case.
+            registerProvider(providers, provider);
+          } else {
+            // Load Avro provider from "core" if it was not loaded from Avro extension before.
+            registerProviderOptionally(providers, provider);
+          }
+        } else {
+          checkState(
+              !providers.containsKey(provider.identifier()),
```
Comment (mosche) [inline sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java:74]: @aromanenko-dev you have to check for `"beam:transform:org.apache.beam:schemaio_" + provider.identifier() + "_read:v1"` or `"beam:transform:org.apache.beam:schemaio_" + provider.identifier() + "_write:v1"`

Comment (aromanenko-dev) [reply sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java:74]: Right-right, copy-paste, sorry... actually, it didn't exist before

**Comment by aromanenko-dev** [comment]
Run Python 3.10 PostCommit

**Comment by aromanenko-dev** [comment]
Run Python 3.10 PostCommit

**Comment by aromanenko-dev** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java [hunk 3]
```diff
@@ -44,19 +48,19 @@ public static <T extends Identifyable> Map<String, T> loadProviders(Class<T> kla
...
-            .getName()
-            .equals(
-                "org.apache.beam.sdk.extensions.avro.schemas.io.payloads.AvroPayloadSerializerProvider")) {
-          // Use AvroPayloadSerializerProvider from extensions/avro by any case.
+        if (provider.getClass().getName().startsWith("org.apache.beam.sdk.extensions.avro")) {
+          // Load Avro provider from "extensions/avro" by any case.
           providers.put(provider.identifier(), provider);
         } else {
           // Load Avro provider from "core" if it was not loaded from Avro extension before.
+          LOG.warn(
```
Comment (mosche) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/Providers.java:60]: Unfortunately this is generating noise and might warn users even if the Avro extension is on the classpath depending on the order :/
I think it's better to keep this as a separate task @aromanenko-dev. Adding warning logs isn't urgent at all.

File: sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java [hunk 4]
```diff
@@ -40,32 +43,75 @@
...
+        // TODO: we won't need this check once all Avro providers from "core" will be
+        // removed
+        if (provider.identifier().equals("avro")) {
+          // Avro provider from "extensions/avro" must have a priority.
+          if (provider.getClass().getName().startsWith("org.apache.beam.sdk.extensions.avro")) {
+            // Load Avro provider from "extensions/avro" by any case.
+            registerProvider(providers, provider);
+          } else {
+            // Load Avro provider from "core" if it was not loaded from Avro extension before.
+            LOG.warn(
```
Comment (mosche) [inline sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java:74]: Same as above regarding logs

**Comment by aromanenko-dev** [comment]
Run Python 3.10 PostCommit

**Comment by mosche** [comment]
Run Python 3.10 PostCommit

**Review by mosche — APPROVED** [review]
👍 This should fix the issue ...

While this adds a bit of code duplication, the actual smell isn't related to this PR in my opinion. Generating (versioned) red / write urns in the registrar seems wrong, these should be defined by the respective provider.

**Comment by damccorm** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by damccorm** [comment]
Run Java_Hadoop_IO_Direct PreCommit

**Comment by damccorm** [comment]
Rerunning flakes impacted by jenkins flakiness (underlying errors were Jenkins errors, not test issues)

**Comment by damccorm** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by damccorm** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by Abacn** [comment]
the affected test passed (seen from log of Python3.10 postcommit):
```
14:30:45 PASSED                                                                   [ 75%]
14:30:45 apache_beam/io/external/xlang_jdbcio_it_test.py::CrossLanguageJdbcIOTest::test_xlang_jdbc_write_read_1_mysql 
14:30:45 -------------------------------- live log call ---------------------------------
14:30:45 INFO     testcontainers.core.container:container.py:53 Pulling image mysql:latest
14:30:47 INFO     testcontainers.core.container:container.py:64 Container started: 22da3809393d
14:30:47 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:47 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:48 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:49 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:50 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:51 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:52 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:53 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:54 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:55 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:56 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:57 INFO     testcontainers.core.waiting_utils:waiting_utils.py:46 Waiting to be ready...
14:30:58 
```

**Pull request merged**